src/selenium_utils.py

```diff
@@ -0,0 +1,55 @@
+from os import EX_PROTOCOL
+from selenium import webdriver
+import selenium
+from selenium.webdriver.firefox.options import Options as Firefox_Options
+from selenium.webdriver.common.keys import Keys
+import logging
+
+#save every page where it errors
+#this helps to progressively catch all the error cases
+def log_error(driver, error_message):
+    logging.error(error_message)
+    driver.take_screenshot("error.png")
```

Comment:
might not be necessary if you recode the html/css/(may not js) into a file "screenshot" using assets
this `logging.debug(get_attribute(driver, "InnerHTML"))` would not be nessary then